File: src/route-engine.mjs
import { rawNodes, rawRoads, rawRoundabouts, rawUTurns } from "./map-data.mjs";

const perf = globalThis.performance ?? { now: () => Date.now() };

function expandPoint(point) {
  return { x: Math.round(point.x * 1.16 + 34), y: Math.round(point.y * 1.12 + 28) };
}

export function edgeKey(a, b) {
  return [a, b].sort().join("-");
}

export function roadWidth(type) {
  return { highway: 52, primary: 44, secondary: 35, local: 28 }[type] || 30;
}

export function createRouteEngine() {
  const nodes = rawNodes.map((node) => ({ ...node, ...expandPoint(node) }));
  const roundabouts = rawRoundabouts.map((roundabout) => {
    const expanded = expandPoint(roundabout);
    return { ...roundabout, ...expanded, r: Math.round(roundabout.r * 1.2) };
  });
  const uTurns = rawUTurns.map((turn) => {
    const expanded = expandPoint(turn);
    return { ...turn, ...expanded, r: Math.round(turn.r * 1.18) };
  });
  const nodeById = new Map(nodes.map((node) => [node.id, node]));
  const adjacency = new Map(nodes.map((node) => [node.id, []]));
  const edgeMap = new Map();
  const roundaboutByNode = new Map();
  let trafficMode = "normal";

  const distance = (a, b) => {
    const from = nodeById.get(a);
    const to = nodeById.get(b);
    return Math.hypot(from.x - to.x, from.y - to.y) / 18;
  };

  rawRoads.forEach(([from, to, traffic, name, type]) => {
    const baseDistance = distance(from, to);
    const edge = { from, to, baseTraffic: traffic, traffic, name, type, distance: baseDistance, cost: baseDistance * traffic };
    edgeMap.set(edgeKey(from, to), edge);
    adjacency.get(from).push({ id: to, from, to, laneSide: -1, traffic, distance: baseDistance, cost: edge.cost });
    adjacency.get(to).push({ id: from, from: to, to: from, laneSide: 1, traffic, distance: baseDistance, cost: edge.cost });
  });

  function nearestNode(point, excludedId = null) {
    return nodes.reduce((nearest, node) => {
      if (node.id === excludedId) return nearest;
      const currentDistance = Math.hypot(node.x - point.x, node.y - point.y);
      return currentDistance < nearest.distance ? { node, distance: currentDistance } : nearest;
    }, { node: nodes[0], distance: Infinity }).node;
  }

  roundabouts.forEach((roundabout) => {
    roundaboutByNode.set(nearestNode(roundabout).id, roundabout);
  });

  const heuristic = (from, to) => distance(from, to);

  function reconstructPath(cameFrom, current) {
    const path = [current];
    while (cameFrom.has(current)) {
      current = cameFrom.get(current);
      path.unshift(current);
    }
    return path;
  }

  function aStar(start, goal) {
    const open = new Set([start]);
    const cameFrom = new Map();
    const gScore = new Map(nodes.map((node) => [node.id, Infinity]));
    const fScore = new Map(nodes.map((node) => [node.id, Infinity]));
    gScore.set(start, 0);
    fScore.set(start, heuristic(start, goal));
    while (open.size > 0) {
      const current = [...open].reduce((best, node) => fScore.get(node) < fScore.get(best) ? node : best);
      if (current === goal) return reconstructPath(cameFrom, current);
      open.delete(current);
      adjacency.get(current).forEach((neighbor) => {
        const tentative = gScore.get(current) + neighbor.cost;
        if (tentative < gScore.get(neighbor.id)) {
          cameFrom.set(neighbor.id, current);
          gScore.set(neighbor.id, tentative);
          fScore.set(neighbor.id, tentative + heuristic(neighbor.id, goal));
          open.add(neighbor.id);
        }
      });
    }
    return [];
  }

  function findAllPaths(start, goal, maxPaths = 14, maxDepth = 9) {
    const paths = [];
    const stack = [[start, [start], 0]];
    const startedAt = perf.now();
    while (stack.length && paths.length < maxPaths) {
      if (perf.now() - startedAt > 45) break;
      const [current, path, cost] = stack.pop();
      if (current === goal) {
        paths.push({ path, cost });
        continue;
      }
      if (path.length >= maxDepth) continue;
      [...adjacency.get(current)]
        .filter((neighbor) => !path.includes(neighbor.id))
        .sort((a, b) => (cost + b.cost + heuristic(b.id, goal)) - (cost + a.cost + heuristic(a.id, goal)))
        .forEach((neighbor) => stack.push([neighbor.id, [...path, neighbor.id], cost + neighbor.cost]));
    }
    return paths.sort((a, b) => a.cost - b.cost);
  }

  const pathCost = (path) => path.slice(0, -1).reduce((sum, id, index) => sum + edgeMap.get(edgeKey(id, path[index + 1])).cost, 0);
  const pathDistance = (path) => path.slice(0, -1).reduce((sum, id, index) => sum + edgeMap.get(edgeKey(id, path[index + 1])).distance, 0);

  function trafficForMode(edge, index) {
    if (trafficMode === "light") return Math.max(1, edge.baseTraffic * 0.72);
    if (trafficMode === "heavy") return Math.min(2.8, edge.baseTraffic * 1.45);
    if (trafficMode === "rush") return Math.min(2.9, Math.max(1, edge.baseTraffic * (0.86 + ((index * 37) % 80) / 50)));
    return edge.baseTraffic;
  }

  function applyTrafficMode(mode = trafficMode) {
    trafficMode = mode;
    [...edgeMap.values()].forEach((edge, index) => {
      edge.traffic = trafficForMode(edge, index);
      edge.cost = edge.distance * edge.traffic;
    });
    adjacency.forEach((neighbors, from) => {
      neighbors.forEach((neighbor) => {
        const edge = edgeMap.get(edgeKey(from, neighbor.id));
        neighbor.traffic = edge.traffic;
        neighbor.cost = edge.cost;
        neighbor.distance = edge.distance;
      });
    });
  }

  function directedLaneSide(fromId, toId) {
    const edge = edgeMap.get(edgeKey(fromId, toId));
    return edge?.from === fromId && edge?.to === toId ? -1 : 1;
  }

  function lanePointOnEdge(edge, t, laneSide) {
    const from = nodeById.get(edge.from);
    const to = nodeById.get(edge.to);
    const dx = to.x - from.x;
    const dy = to.y - from.y;
    const length = Math.hypot(dx, dy) || 1;
    const nx = -dy / length;
    const ny = dx / length;
    const laneOffset = Math.max(7, roadWidth(edge.type) * 0.24);
    return { x: from.x + dx * t + nx * laneOffset * laneSide, y: from.y + dy * t + ny * laneOffset * laneSide };
  }

  const travelTForNode = (edge, nodeId) => edge.from === nodeId ? 0 : 1;

  function roundaboutTrimmedT(edge, nodeId) {
    const roundabout = roundaboutByNode.get(nodeId);
    if (!roundabout) return travelTForNode(edge, nodeId);
    const from = nodeById.get(edge.from);
    const to = nodeById.get(edge.to);
    const length = Math.hypot(to.x - from.x, to.y - from.y) || 1;
    const trim = Math.min(0.35, (roundabout.r + roadWidth(edge.type) * 0.08) / length);
    return edge.from === nodeId ? trim : 1 - trim;
  }

  function lanePointForDirectedNode(nodeId, fromId, toId) {
    const edge = edgeMap.get(edgeKey(fromId, toId));
    if (!edge) return nodeById.get(nodeId);
    return lanePointOnEdge(edge, roundaboutTrimmedT(edge, nodeId), directedLaneSide(fromId, toId));
  }

  function sampleDirectedEdge(fromId, toId, startT = null, endT = null) {
    const edge = edgeMap.get(edgeKey(fromId, toId));
    if (!edge) return [];
    const fromT = startT ?? roundaboutTrimmedT(edge, fromId);
    const toT = endT ?? roundaboutTrimmedT(edge, toId);
    const steps = Math.max(4, Math.ceil(Math.abs(toT - fromT) * edge.distance / 2.4));
    return Array.from({ length: steps + 1 }, (_, i) => lanePointOnEdge(edge, fromT + (toT - fromT) * (i / steps), directedLaneSide(fromId, toId)));
  }

  function tForPointOnEdge(point, edge) {
    const from = nodeById.get(edge.from);
    const to = nodeById.get(edge.to);
    const dx = to.x - from.x;
    const dy = to.y - from.y;
    return Math.max(0, Math.min(1, ((point.x - from.x) * dx + (point.y - from.y) * dy) / (dx * dx + dy * dy || 1)));
  }

  function nearestRoadPoint(point, excludedId = null) {
    return [...edgeMap.values()].reduce((nearest, edge) => {
      const from = nodeById.get(edge.from);
      const to = nodeById.get(edge.to);
      const dx = to.x - from.x;
      const dy = to.y - from.y;
      const lengthSquared = dx * dx + dy * dy || 1;
      const length = Math.sqrt(lengthSquared);
      const nx = -dy / length;
      const ny = dx / length;
      const t = Math.max(0, Math.min(1, ((point.x - from.x) * dx + (point.y - from.y) * dy) / lengthSquared));
      const center = { x: from.x + dx * t, y: from.y + dy * t };
      const signedDistance = (point.x - center.x) * nx + (point.y - center.y) * ny;
      const laneSide = signedDistance < 0 ? -1 : 1;
      const laneStartNode = laneSide === -1 ? edge.from : edge.to;
      const laneEndNode = laneSide === -1 ? edge.to : edge.from;
      const graphNode = laneStartNode === excludedId ? laneEndNode : laneStartNode;
      const distanceToRoad = Math.abs(signedDistance);
      return distanceToRoad < nearest.distance
        ? { point: lanePointOnEdge(edge, t, laneSide), graphNode, laneStartNode, laneEndNode, laneT: t, distance: distanceToRoad, edge, laneSide }
        : nearest;
    }, { point, graphNode: nearestNode(point, excludedId).id, laneStartNode: null, laneEndNode: null, laneT: 0, distance: Infinity, edge: null, laneSide: 1 });
  }

  function angleBetweenPoints(from, to) {
    return Math.atan2(to.y - from.y, to.x - from.x);
  }

  function normalizeAngle(angle) {
    while (angle < 0) angle += Math.PI * 2;
    while (angle >= Math.PI * 2) angle -= Math.PI * 2;
    return angle;
  }

  function sampleRoundaboutArc(roundabout, entryPoint, exitPoint) {
    const start = normalizeAngle(angleBetweenPoints(roundabout, entryPoint));
    const end = normalizeAngle(angleBetweenPoints(roundabout, exitPoint));
    let delta = end - start;
    if (delta < 0) delta += Math.PI * 2;
    if (delta > Math.PI * 1.35) delta -= Math.PI * 2;
    const steps = Math.max(5, Math.ceil(Math.abs(delta) / (Math.PI / 8)));
    return Array.from({ length: steps + 1 }, (_, i) => {
      const angle = start + delta * (i / steps);
      return { x: roundabout.x + Math.cos(angle) * roundabout.r, y: roundabout.y + Math.sin(angle) * roundabout.r };
    });
  }

  const samePoint = (a, b) => a && b && Math.hypot(a.x - b.x, a.y - b.y) < 0.5;
  const pushPoint = (points, point) => { if (point && !samePoint(points[points.length - 1], point)) points.push(point); };

  function pathToPoints(path, startPin, endPin, startPoint, endPoint) {
    if (!path.length) return [];
    const routedPoints = [];
    if (startPin?.edge) {
      sampleDirectedEdge(startPin.laneStartNode, startPin.laneEndNode, tForPointOnEdge(startPoint, startPin.edge), null).forEach((point) => pushPoint(routedPoints, point));
    } else {
      pushPoint(routedPoints, startPoint);
    }
    path.slice(0, -1).forEach((fromId, index) => {
      const toId = path[index + 1];
      const previousPoint = routedPoints[routedPoints.length - 1];
      const nextStart = lanePointForDirectedNode(fromId, fromId, toId);
      const roundabout = roundaboutByNode.get(fromId);
      if (roundabout && previousPoint && nextStart && !samePoint(previousPoint, nextStart)) {
        sampleRoundaboutArc(roundabout, previousPoint, nextStart).forEach((point) => pushPoint(routedPoints, point));
      } else {
        pushPoint(routedPoints, nextStart);
      }
      sampleDirectedEdge(fromId, toId).forEach((point) => pushPoint(routedPoints, point));
    });
    if (endPin?.edge) {
      const finalNode = path[path.length - 1];
      const endEdgeStart = lanePointForDirectedNode(finalNode, endPin.laneStartNode, endPin.laneEndNode);
      const roundabout = roundaboutByNode.get(finalNode);
      if (roundabout && !samePoint(routedPoints[routedPoints.length - 1], endEdgeStart)) {
        sampleRoundaboutArc(roundabout, routedPoints[routedPoints.length - 1], endEdgeStart).forEach((point) => pushPoint(routedPoints, point));
      } else {
        pushPoint(routedPoints, endEdgeStart);
      }
      sampleDirectedEdge(endPin.laneStartNode, endPin.laneEndNode, null, tForPointOnEdge(endPoint, endPin.edge)).forEach((point) => pushPoint(routedPoints, point));
    } else {
      pushPoint(routedPoints, endPoint);
    }
    return routedPoints;
  }

  applyTrafficMode("normal");
  return { nodes, nodeById, adjacency, edgeMap, roundabouts, uTurns, roundaboutByNode, aStar, findAllPaths, pathCost, pathDistance, applyTrafficMode, nearestRoadPoint, pathToPoints, roadWidth };
}

File: python/route_engine.py
import heapq
import math
import time
from .map_data import RAW_NODES, RAW_ROADS, RAW_ROUNDABOUTS


ROAD_WIDTHS = {"highway": 52, "primary": 44, "secondary": 35, "local": 28}


def expand_point(point):
    return {
        **point,
        "x": round(point["x"] * 1.16 + 34),
        "y": round(point["y"] * 1.12 + 28),
    }


def edge_key(a, b):
    return "-".join(sorted((a, b)))


def road_width(road_type):
    return ROAD_WIDTHS.get(road_type, 30)


def traffic_for_mode(base_traffic, index, mode):
    if mode == "light":
        return max(1, base_traffic * 0.72)
    if mode == "heavy":
        return min(2.8, base_traffic * 1.45)
    if mode == "rush":
        wave = 0.86 + ((index * 37) % 80) / 50
        return min(2.9, max(1, base_traffic * wave))
    return base_traffic


class RouteEngine:
    def __init__(self, traffic_mode="normal"):
        self.nodes = [expand_point(node) for node in RAW_NODES]
        self.roundabouts = [
            {**roundabout, **expand_point(roundabout), "r": round(roundabout["r"] * 1.2)}
            for roundabout in RAW_ROUNDABOUTS
        ]
        self.node_by_id = {node["id"]: node for node in self.nodes}
        self.adjacency = {node["id"]: [] for node in self.nodes}
        self.edge_map = {}
        self.roundabout_by_node = {}
        self.traffic_mode = traffic_mode
        self._build_graph()
        for roundabout in self.roundabouts:
            self.roundabout_by_node[self.nearest_node(roundabout)["id"]] = roundabout

    def distance(self, a, b):
        from_node = self.node_by_id[a]
        to_node = self.node_by_id[b]
        return math.hypot(from_node["x"] - to_node["x"], from_node["y"] - to_node["y"]) / 18

    def _build_graph(self):
        self.adjacency = {node["id"]: [] for node in self.nodes}
        self.edge_map = {}
        for index, (from_id, to_id, base_traffic, name, road_type) in enumerate(RAW_ROADS):
            distance = self.distance(from_id, to_id)
            traffic = traffic_for_mode(base_traffic, index, self.traffic_mode)
            edge = {
                "from": from_id,
                "to": to_id,
                "baseTraffic": base_traffic,
                "traffic": traffic,
                "name": name,
                "type": road_type,
                "distance": distance,
                "cost": distance * traffic,
            }
            self.edge_map[edge_key(from_id, to_id)] = edge
            self.adjacency[from_id].append({
                "id": to_id,
                "from": from_id,
                "to": to_id,
                "laneSide": -1,
                "cost": edge["cost"],
                "distance": distance,
                "traffic": traffic,
            })
            self.adjacency[to_id].append({
                "id": from_id,
                "from": to_id,
                "to": from_id,
                "laneSide": 1,
                "cost": edge["cost"],
                "distance": distance,
                "traffic": traffic,
            })

    def nearest_node(self, point, excluded_id=None):
        nearest = None
        nearest_distance = math.inf
        for node in self.nodes:
            if node["id"] == excluded_id:
                continue
            current_distance = math.hypot(node["x"] - point["x"], node["y"] - point["y"])
            if current_distance < nearest_distance:
                nearest = node
                nearest_distance = current_distance
        return nearest

    def directed_lane_side(self, from_id, to_id):
        edge = self.edge_map.get(edge_key(from_id, to_id))
        return -1 if edge and edge["from"] == from_id and edge["to"] == to_id else 1

    def lane_point_on_edge(self, edge, t, lane_side):
        from_node = self.node_by_id[edge["from"]]
        to_node = self.node_by_id[edge["to"]]
        dx = to_node["x"] - from_node["x"]
        dy = to_node["y"] - from_node["y"]
        length = math.hypot(dx, dy) or 1
        nx = -dy / length
        ny = dx / length
        lane_offset = max(7, road_width(edge["type"]) * 0.24)
        return {
            "x": from_node["x"] + dx * t + nx * lane_offset * lane_side,
            "y": from_node["y"] + dy * t + ny * lane_offset * lane_side,
        }

    @staticmethod
    def travel_t_for_node(edge, node_id):
        return 0 if edge["from"] == node_id else 1

    def roundabout_trimmed_t(self, edge, node_id):
        roundabout = self.roundabout_by_node.get(node_id)
        if not roundabout:
            return self.travel_t_for_node(edge, node_id)
        from_node = self.node_by_id[edge["from"]]
        to_node = self.node_by_id[edge["to"]]
        length = math.hypot(to_node["x"] - from_node["x"], to_node["y"] - from_node["y"]) or 1
        trim = min(0.35, (roundabout["r"] + road_width(edge["type"]) * 0.08) / length)
        return trim if edge["from"] == node_id else 1 - trim

    def lane_point_for_directed_node(self, node_id, from_id, to_id):
        edge = self.edge_map.get(edge_key(from_id, to_id))
        if not edge:
            return self.node_by_id[node_id]
        return self.lane_point_on_edge(
            edge,
            self.roundabout_trimmed_t(edge, node_id),
            self.directed_lane_side(from_id, to_id),
        )

    def sample_directed_edge(self, from_id, to_id, start_t=None, end_t=None):
        edge = self.edge_map.get(edge_key(from_id, to_id))
        if not edge:
            return []
        from_t = self.roundabout_trimmed_t(edge, from_id) if start_t is None else start_t
        to_t = self.roundabout_trimmed_t(edge, to_id) if end_t is None else end_t
        steps = max(4, math.ceil(abs(to_t - from_t) * edge["distance"] / 2.4))
        return [
            self.lane_point_on_edge(
                edge,
                from_t + (to_t - from_t) * (index / steps),
                self.directed_lane_side(from_id, to_id),
            )
            for index in range(steps + 1)
        ]

    def t_for_point_on_edge(self, point, edge):
        from_node = self.node_by_id[edge["from"]]
        to_node = self.node_by_id[edge["to"]]
        dx = to_node["x"] - from_node["x"]
        dy = to_node["y"] - from_node["y"]
        denominator = dx * dx + dy * dy or 1
        return max(0, min(1, ((point["x"] - from_node["x"]) * dx + (point["y"] - from_node["y"]) * dy) / denominator))

    def nearest_road_point(self, point, excluded_id=None):
        nearest = {
            "point": point,
            "graphNode": self.nearest_node(point, excluded_id)["id"],
            "laneStartNode": None,
            "laneEndNode": None,
            "laneT": 0,
            "distance": math.inf,
            "edge": None,
            "laneSide": 1,
        }
        for edge in self.edge_map.values():
            from_node = self.node_by_id[edge["from"]]
            to_node = self.node_by_id[edge["to"]]
            dx = to_node["x"] - from_node["x"]
            dy = to_node["y"] - from_node["y"]
            length_squared = dx * dx + dy * dy or 1
            t = max(0, min(1, ((point["x"] - from_node["x"]) * dx + (point["y"] - from_node["y"]) * dy) / length_squared))
            candidates = []
            for lane_side in (-1, 1):
                lane_point = self.lane_point_on_edge(edge, t, lane_side)
                candidates.append({
                    "laneSide": lane_side,
                    "lanePoint": lane_point,
                    "distance": math.hypot(point["x"] - lane_point["x"], point["y"] - lane_point["y"]),
                })
            closest = candidates[0] if candidates[0]["distance"] <= candidates[1]["distance"] else candidates[1]
            lane_side = closest["laneSide"]
            lane_start_node = edge["from"] if lane_side == -1 else edge["to"]
            lane_end_node = edge["to"] if lane_side == -1 else edge["from"]
            graph_node = lane_end_node if lane_start_node == excluded_id else lane_start_node
            if closest["distance"] < nearest["distance"]:
                nearest = {
                    "point": closest["lanePoint"],
                    "graphNode": graph_node,
                    "laneStartNode": lane_start_node,
                    "laneEndNode": lane_end_node,
                    "laneT": t,
                    "distance": closest["distance"],
                    "edge": edge,
                    "laneSide": lane_side,
                }
        return nearest

    @staticmethod
    def angle_between_points(from_point, to_point):
        return math.atan2(to_point["y"] - from_point["y"], to_point["x"] - from_point["x"])

    @staticmethod
    def normalize_angle(angle):
        while angle < 0:
            angle += math.pi * 2
        while angle >= math.pi * 2:
            angle -= math.pi * 2
        return angle

    def sample_roundabout_arc(self, roundabout, entry_point, exit_point):
        start = self.normalize_angle(self.angle_between_points(roundabout, entry_point))
        end = self.normalize_angle(self.angle_between_points(roundabout, exit_point))
        delta = end - start
        if delta < 0:
            delta += math.pi * 2
        if delta > math.pi * 1.35:
            delta -= math.pi * 2
        steps = max(5, math.ceil(abs(delta) / (math.pi / 8)))
        return [
            {
                "x": roundabout["x"] + math.cos(start + delta * (index / steps)) * roundabout["r"],
                "y": roundabout["y"] + math.sin(start + delta * (index / steps)) * roundabout["r"],
            }
            for index in range(steps + 1)
        ]

    @staticmethod
    def same_point(a, b):
        return bool(a and b and math.hypot(a["x"] - b["x"], a["y"] - b["y"]) < 0.5)

    def push_point(self, points, point):
        if point and not self.same_point(points[-1] if points else None, point):
            points.append(point)

    def path_to_points(self, path, start_pin=None, end_pin=None, start_point=None, end_point=None):
        if not path:
            return []
        routed_points = []
        if start_pin and start_pin.get("edge"):
            self.push_point(routed_points, start_pin["point"])
            for point in self.sample_directed_edge(
                start_pin["laneStartNode"],
                start_pin["laneEndNode"],
                self.t_for_point_on_edge(start_point or start_pin["point"], start_pin["edge"]),
                None,
            ):
                self.push_point(routed_points, point)
        elif start_point:
            self.push_point(routed_points, start_point)

        for index, from_id in enumerate(path[:-1]):
            to_id = path[index + 1]
            previous_point = routed_points[-1] if routed_points else None
            next_start = self.lane_point_for_directed_node(from_id, from_id, to_id)
            roundabout = self.roundabout_by_node.get(from_id)
            if roundabout and previous_point and next_start and not self.same_point(previous_point, next_start):
                for point in self.sample_roundabout_arc(roundabout, previous_point, next_start):
                    self.push_point(routed_points, point)
            else:
                self.push_point(routed_points, next_start)
            for point in self.sample_directed_edge(from_id, to_id):
                self.push_point(routed_points, point)

        if end_pin and end_pin.get("edge"):
            final_node = path[-1]
            end_edge_start = self.lane_point_for_directed_node(final_node, end_pin["laneStartNode"], end_pin["laneEndNode"])
            roundabout = self.roundabout_by_node.get(final_node)
            if roundabout and routed_points and not self.same_point(routed_points[-1], end_edge_start):
                for point in self.sample_roundabout_arc(roundabout, routed_points[-1], end_edge_start):
                    self.push_point(routed_points, point)
            else:
                self.push_point(routed_points, end_edge_start)
            for point in self.sample_directed_edge(
                end_pin["laneStartNode"],
                end_pin["laneEndNode"],
                None,
                self.t_for_point_on_edge(end_point or end_pin["point"], end_pin["edge"]),
            ):
                self.push_point(routed_points, point)
        elif end_point:
            self.push_point(routed_points, end_point)

        return routed_points

    def heuristic(self, start, goal):
        return self.distance(start, goal)

    def a_star(self, start, goal):
        open_heap = [(self.heuristic(start, goal), start)]
        came_from = {}
        g_score = {node["id"]: math.inf for node in self.nodes}
        g_score[start] = 0
        visited = set()

        while open_heap:
            _, current = heapq.heappop(open_heap)
            if current in visited:
                continue
            if current == goal:
                return self.reconstruct_path(came_from, current)
            visited.add(current)

            for neighbor in self.adjacency[current]:
                tentative = g_score[current] + neighbor["cost"]
                if tentative < g_score[neighbor["id"]]:
                    came_from[neighbor["id"]] = current
                    g_score[neighbor["id"]] = tentative
                    heapq.heappush(open_heap, (tentative + self.heuristic(neighbor["id"], goal), neighbor["id"]))

        return []

    @staticmethod
    def reconstruct_path(came_from, current):
        path = [current]
        while current in came_from:
            current = came_from[current]
            path.insert(0, current)
        return path

    def find_all_paths(self, start, goal, max_paths=14, max_depth=9):
        paths = []
        stack = [(start, [start], 0)]
        started_at = time.perf_counter()
        while stack and len(paths) < max_paths:
            if (time.perf_counter() - started_at) * 1000 > 45:
                break
            current, path, cost = stack.pop()
            if current == goal:
                paths.append({"path": path, "cost": cost})
                continue
            if len(path) >= max_depth:
                continue
            neighbors = [neighbor for neighbor in self.adjacency[current] if neighbor["id"] not in path]
            neighbors.sort(key=lambda n: cost + n["cost"] + self.heuristic(n["id"], goal), reverse=True)
            for neighbor in neighbors:
                stack.append((neighbor["id"], path + [neighbor["id"]], cost + neighbor["cost"]))
        return sorted(paths, key=lambda item: item["cost"])

    def path_cost(self, path):
        return sum(self.edge_map[edge_key(path[i], path[i + 1])]["cost"] for i in range(len(path) - 1))

    def path_distance(self, path):
        return sum(self.edge_map[edge_key(path[i], path[i + 1])]["distance"] for i in range(len(path) - 1))

    def routes(self, start, end):
        best_path = self.a_star(start, end)
        possible_paths = self.find_all_paths(start, end)
        return {
            "bestPath": best_path,
            "possiblePaths": possible_paths,
            "cost": self.path_cost(best_path) if best_path else 0,
            "distance": self.path_distance(best_path) if best_path else 0,
            "trafficMode": self.traffic_mode,
            "source": "python",
        }


def calculate_routes(start, end, traffic_mode="normal"):
    return RouteEngine(traffic_mode).routes(start, end)
